synapse/storage/databases/main/roommember.py

```diff
@@ -488,13 +488,16 @@ async def get_rooms_for_user(self, user_id: str, on_invalidate=None):
 
     @cached(max_entries=500000, cache_context=True, iterable=True)
     async def get_users_who_share_room_with_user(
-        self, user_id: str, cache_context: _CacheContext
+        self, user_id: str, encrypted_only: bool, cache_context: _CacheContext
     ) -> Set[str]:
         """Returns the set of users who share a room with `user_id`"""
         room_ids = await self.get_rooms_for_user(
             user_id, on_invalidate=cache_context.invalidate
         )
 
+        if encrypted_only:
+            room_ids = await self._filter_encrypted_rooms(room_ids)
```

Comment:
We'd need to invalidate the cached result if the encryption state of the room change. I think we'd probably want to change `_filter_encrypted_rooms` to be a `get_encryption_state_for_rooms` and add a cache on that, so that we can `on_invalidate=cache_context.invalidate` when we call it, and so it'll "do the right thing".